### PR TITLE
feat: Introduce CamelK extension

### PR DESCRIPTION
**Pull request by iocanel**
-------

## Type of change

 - [ ] Bug fix (non-breaking change which fixes an issue)
 - [x] Feature (non-breaking change which adds functionality)
 - [ ] Breaking change (fix or feature that would cause existing functionality to change
 - [ ] Chore (non-breaking change which doesn't affect codebase;
   test, version modification, documentation, etc.)

## Checklist
 - [ ] Code contributed by me aligns with current project license: [Apache 2.0](https://www.apache.org/licenses/LICENSE-2.0)
 - [x] I Added [CHANGELOG](../CHANGELOG.md) entry regarding this change
 - [x] I have implemented unit tests to cover my changes
 - [ ] I have added/updated the [javadocs](https://www.javadoc.io/doc/io.fabric8/kubernetes-client/latest/index.html) and other [documentation](https://github.com/fabric8io/kubernetes-client/blob/master/doc/CHEATSHEET.md) accordingly
 - [ ] No new bugs, code smells, etc. in [SonarCloud](https://sonarcloud.io/dashboard?id=fabric8io_kubernetes-client) report
 - [ ] I tested my code in Kubernetes
 - [ ] I tested my code in OpenShift

### CODE REVIEW

**Comment by lburgazzoli** [comment]
@iocanel create a [PR](https://github.com/ioforks/kubernetes-client/pull/1) against your fork to fix some of the issues

**Comment by iocanel** [comment]
> @iocanel create a [PR](https://github.com/ioforks/kubernetes-client/pull/1) against your fork to fix some of the issues

Merged!

**Comment by lburgazzoli** [comment]
@iocanel I think this branch need to be rebased (in the meantime, version got bump to 5.3-SNAPSHOT)

**Comment by dhirajsb** [comment]
@lburgazzoli I'm rebasing and fixing the build issues.

**Comment by dhirajsb** [comment]
@iocanel fixed build issues with [PR](https://github.com/ioforks/kubernetes-client/pull/2/), please merge

**Comment by iocanel** [comment]
> @iocanel fixed build issues with [PR](https://github.com/ioforks/kubernetes-client/pull/2/), please merge

@dhirajsb: Thanks a lot! Merged!

**Comment by iocanel** [comment]
@dhirajsb: It seems that all related build issues are now fixed.

Once reviewed, I think that we can merged that!

**Comment by lburgazzoli** [comment]
@iocanel @dhirajsb I've built the extension locally and the Flow is expressed as "java.lang.String", I think it should be mapped to a a JsonNode otherwise de.serializaion would fail as in the CR, it is expressed a Json object.

**Comment by iocanel** [comment]
> @iocanel @dhirajsb I've built the extension locally and the Flow is expressed as "java.lang.String", I think it should be mapped to a a JsonNode otherwise de.serializaion would fail as in the CR, it is expressed a Json object.
> 
> I forgot to update also the v1alpha generator with:
> 
> ```
> {GoType: reflect.TypeOf(v1.Flow{}), JavaClass: "com.fasterxml.jackson.databind.JsonNode"}
> ```

My guess is that after you updated the generator with the correct mappings, you didn't run `make` or you just didn't commit the genrated json descriptors.

File: CHANGELOG.md [hunk 1]
```diff
@@ -41,6 +41,10 @@
 * Fix #2699: Add support for `@EnableOpenShiftMockClient` for OpenShiftClient
 * Fix #2588: Support for Server side dry runs
 
+- CamelK extension
```
Comment (metacosm) [inline CHANGELOG.md:44]: Please reference the associated issue, using the format of the other changelog entries.

File: CHANGELOG.md [hunk 2]
```diff
@@ -41,6 +41,10 @@
...
+
+- CamelK extension
```
Comment (metacosm) [inline CHANGELOG.md:46]:
```suggestion
```

File: extensions/camel-k/client/src/main/java/io/fabric8/camelk/client/dsl/V1APIGroupDSL.java [hunk 3]
```diff
@@ -0,0 +1,70 @@
...
+import io.fabric8.kubernetes.client.dsl.NonNamespaceOperation;
+import io.fabric8.kubernetes.client.dsl.Resource;
+import io.fabric8.camelk.v1.Build;
+import io.fabric8.camelk.v1.BuildList;
+import io.fabric8.camelk.v1.CamelCatalog;
+import io.fabric8.camelk.v1.CamelCatalogList;
+import io.fabric8.camelk.v1.Integration;
+import io.fabric8.camelk.v1.IntegrationList;
+import io.fabric8.camelk.v1.IntegrationKit;
+import io.fabric8.camelk.v1.IntegrationKitList;
```
Comment (metacosm) [inline extensions/camel-k/client/src/main/java/io/fabric8/camelk/client/dsl/V1APIGroupDSL.java:29]: Do we really need to keep generating List objects?

File: extensions/camel-k/client/src/main/java/io/fabric8/camelk/client/util/ApiVersionUtil.java [hunk 4]
```diff
@@ -0,0 +1,103 @@
...
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package io.fabric8.camelk.client.util;
+
+import io.fabric8.kubernetes.api.model.HasMetadata;
+import io.fabric8.kubernetes.client.utils.Utils;
+
+/** Borrowed from the client. */
+public class ApiVersionUtil {
```
Comment (metacosm) [inline extensions/camel-k/client/src/main/java/io/fabric8/camelk/client/util/ApiVersionUtil.java:22]: This should probably be moved outside of the client project indeed at some point… :(

**Comment by lburgazzoli** [comment]
let me try to re-generate

**Review by metacosm — COMMENTED** [review]
LGTM from a quick glance but I don't think I'm competent to review this PR adequately since I know nothing of Camel-K… :(

**Comment by lburgazzoli** [comment]
> My guess is that after you updated the generator with the correct mappings, you didn't run `make` or you just didn't commit the genrated json descriptors.

Done with https://github.com/ioforks/kubernetes-client/pull/3

**Comment by lburgazzoli** [comment]
@iocanel I think we also need to include x-kubernetes-preserve-unknown-fields for the Flow and other stuffs to work, is there a way to add it ?

**Comment by lburgazzoli** [comment]
I've noticed that some CRDs YAMLs are not properly generated, as example, see the extract from the `KameletBinding` below: 

```yaml
binding:
  properties:
    sink:
      properties:
        properties:
          type: object
        ref:
          type: object
        types:
          type: object
        uri:
          type: string
        additionalProperties:
          type: object
      type: object
```

Here `ref` is defined as generic `object` without any properties but it does not seem correct as `ref` is an `ObjectReferece` so I'd expect to have the related properties listed otherwise the serialization/de-serialization will ignore anything set on such element as all the fields are unknown.

**Comment by dhirajsb** [comment]
@iocanel the gpg `passphrase` needs to be moved out of the pom.xml for the current failure. Do you know where it can be added in the CI build as perhaps an environment variable?

File: CHANGELOG.md (hunk 1 above)
Comment (dhirajsb) [reply CHANGELOG.md:44]: https://github.com/ioforks/kubernetes-client/pull/4

File: CHANGELOG.md (hunk 2 above)
Comment (dhirajsb) [reply CHANGELOG.md:46]: https://github.com/ioforks/kubernetes-client/pull/4

**Comment by dhirajsb** [comment]
@iocanel please merge https://github.com/ioforks/kubernetes-client/pull/4

File: extensions/camel-k/client/src/main/java/io/fabric8/camelk/client/util/ApiVersionUtil.java (hunk 4 above)
Comment (manusa) [reply extensions/camel-k/client/src/main/java/io/fabric8/camelk/client/util/ApiVersionUtil.java:22]: Maybe move this to the kubernetes-model-common module?

File: extensions/camel-k/model-v1/src/test/java/io/fabric8/camelk/ModelTest.java [hunk 5]
```diff
@@ -0,0 +1,36 @@
...
+package io.fabric8.camelk;
+
+import io.fabric8.camelk.v1.Integration;
+import io.fabric8.camelk.v1.IntegrationBuilder;
+import org.junit.jupiter.api.Test;
+
+class ModelTest {
+
+  @Test
+  void shouldCreateIntegration()  {
```
Comment (manusa) [inline extensions/camel-k/model-v1/src/test/java/io/fabric8/camelk/ModelTest.java:25]: This test is missing assertions

File: extensions/camel-k/model-v1alpha1/pom.xml [hunk 6]
```diff
@@ -0,0 +1,160 @@
...
+      </dependency>
+      <dependency>
+        <groupId>io.fabric8</groupId>
+        <artifactId>knative-model</artifactId>
+      </dependency>
+      <dependency>
+        <groupId>io.fabric8</groupId>
+        <artifactId>camel-k-model-v1</artifactId>
+        <version>${project.version}</version>
+      </dependency>
```
Comment (manusa) [inline extensions/camel-k/model-v1alpha1/pom.xml:71]: This dependency seems odd, why does v1alpha1 depend on v1?

Comment (lburgazzoli) [reply extensions/camel-k/model-v1alpha1/pom.xml:71]: in camel-k we have some types that have matured from v1alpha to v1, but some CRs have not yet moved to v1

Comment (manusa) [reply extensions/camel-k/model-v1alpha1/pom.xml:71]: Yes, but I read this as the opposite. According to your statement, shouldn't v1 depend on v1alpha1?

Comment (lburgazzoli) [reply extensions/camel-k/model-v1alpha1/pom.xml:71]: I think my explanation was not really clear :)

The Flow type is part of v1 but the KameletBinding which is part of v1alpha1 uses Flow hence, v1alpha1 depends on v1

Comment (manusa) [reply extensions/camel-k/model-v1alpha1/pom.xml:71]: :exploding_head:
All clear then :)

**Review by manusa — APPROVED** [review]
LGTM, thx!

**Comment by iocanel** [comment]
> @iocanel please merge [ioforks#4](https://github.com/ioforks/kubernetes-client/pull/4)

Merged! Sorry for the delay!

**Comment by iocanel** [comment]
@lburgazzoli @dhirajsb: I am looking at the commit log of this pull request and it seems a little bit weird to me. In particular my inital commit seems to be present twice. My git foo is not strong enough to understand how it did happen (I would guess that at some point merges  instead of rebasing was used?). Either-way, please do have a look just to make sure none of your changes were somehow got overwritten.

File: extensions/camel-k/client/src/main/java/io/fabric8/camelk/client/CamelKClient.java [hunk 7]
```diff
@@ -0,0 +1,39 @@
...
+import io.fabric8.kubernetes.client.Client;
+import io.fabric8.camelk.client.dsl.V1alpha1APIGroupDSL;
+import io.fabric8.camelk.client.dsl.V1APIGroupDSL;
+
+/**
+ * Main interface for CamelK client library.
+ */
+public interface CamelKClient extends Client {
+  /**
+   * API entrypoint for tekton.dev/v1beta1 API group resources
```
Comment (rohanKanojia) [inline extensions/camel-k/client/src/main/java/io/fabric8/camelk/client/CamelKClient.java:27]: I think you've got apigroup + version wrong here. Maybe it should be `camel.apache.org/v1`

File: extensions/camel-k/client/src/main/java/io/fabric8/camelk/client/CamelKClient.java [hunk 8]
```diff
@@ -0,0 +1,39 @@
...
+   *
+   * @return {@link V1APIGroupDSL} for CamelK resource operations in this API group.
+   */
+  V1APIGroupDSL v1();
+
+  /**
+   * API entrypoint for tekton.dev/v1alpha1 API group resources
```
Comment (rohanKanojia) [inline extensions/camel-k/client/src/main/java/io/fabric8/camelk/client/CamelKClient.java:34]: https://github.com/fabric8io/kubernetes-client/pull/2891/files#r599456186

**Review by oscerd — APPROVED** [review]

File: extensions/camel-k/generator-v1/cmd/generate/generate.go [hunk 9]
```diff
@@ -0,0 +1,88 @@
...
+		reflect.TypeOf(v1.BuildList{}):               schemagen.Namespaced,
+		reflect.TypeOf(v1.CamelCatalogList{}):        schemagen.Namespaced,
+		reflect.TypeOf(v1.IntegrationList{}):         schemagen.Namespaced,
+		reflect.TypeOf(v1.IntegrationKitList{}):      schemagen.Namespaced,
+		reflect.TypeOf(v1.IntegrationPlatformList{}): schemagen.Namespaced,
+	}
+
+	// constraints and patterns for fields
+	constraints := map[reflect.Type]map[string]*schemagen.Constraint{
+		//		reflect.TypeOf(v1.Step{}): {"Name": &schemagen.Constraint{MaxLength: 63, Pattern: "^[a-z0-9]([-a-z0-9]*[a-z0-9])?$"}},
```
Comment (rohanKanojia) [inline extensions/camel-k/generator-v1/cmd/generate/generate.go:43]: Please remove these commented out lines if not required

File: extensions/camel-k/generator-v1/go.mod [hunk 10]
```diff
@@ -0,0 +1,19 @@
+module github.com/fabric8io/kubernetes-client/extensions/camel-k/generator
+
+require (
+	github.com/apache/camel-k v1.3.1 // indirect
+	github.com/apache/camel-k/pkg/apis/camel v1.3.1
+	github.com/fabric8io/kubernetes-client/generator v0.0.0
+	github.com/tektoncd/pipeline v0.13.1-0.20200625065359-44f22a067b75 // indirect
```
Comment (rohanKanojia) [inline extensions/camel-k/generator-v1/go.mod:7]: Is this really required?

File: extensions/camel-k/client/src/main/java/io/fabric8/camelk/client/V1APIGroupClient.java [hunk 11]
```diff
@@ -0,0 +1,75 @@
...
+  public V1APIGroupClient() {
+    super();
+  }
+
+  public V1APIGroupClient(OkHttpClient httpClient, final Config config) {
+    super(httpClient, config);
+  }
+
+@Override
+public MixedOperation<Build, BuildList, Resource<Build>> builds() {
```
Comment (rohanKanojia) [inline extensions/camel-k/client/src/main/java/io/fabric8/camelk/client/V1APIGroupClient.java:51]: nit, indentation seems slightly off here

File: extensions/camel-k/model-v1/pom.xml [hunk 12]
```diff
@@ -0,0 +1,150 @@
...
+      <groupId>io.sundr</groupId>
+      <artifactId>transform-annotations</artifactId>
+    </dependency>
+    <dependency>
+      <groupId>io.fabric8</groupId>
+      <artifactId>kubernetes-model-core</artifactId>
+    </dependency>
+    <dependency>
+      <groupId>io.fabric8</groupId>
+      <artifactId>knative-model</artifactId>
```
Comment (rohanKanojia) [inline extensions/camel-k/model-v1/pom.xml:63]: Does camel k model depend on `knative-model`? I removed this dependency but still was able to see the build being successful.

**Comment by manusa** [comment]
We were waiting to see if we could merge this PR before the rest, since it is involved with many files that will get conflicts after other PR merges, but the rest are piling up.
Please check if you can organize the commit history or confirm if squashing would be OK so we can move on to the rest.

**Comment by manusa** [comment]
I fixed merged conflicts and sorted out commit history in #2966, please review and check I didn't lose anything in the process so that we can merge today.

**Pull request closed without merging**